### ISSUE TITLE
Regression in clang build?

### ISSUE THREAD

**Issue by vchong** [OP-TEE/optee_os#3762]
On latest master (aeb5ba43), the clang build doesn't seem to work for hikey960? Shippable isn't showing any errors, although there are warnings like below, so not sure if it's user error.

```
_make PLATFORM=hikey-hikey960 CFG_ARM64_core=y COMPILER=clang
llvm-objdump: warning: 'out/ldelf/ldelf.elf': failed to parse debug information for out/ldelf/ldelf.elf
llvm-objdump: warning: 'out/ta/avb/023f8f1a-292a-432b-8fc4-de8471358067.elf': failed to parse debug information for out/ta/avb/023f8f1a-292a-432b-8fc4-de8471358067.elf
llvm-objdump: warning: 'out/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.elf': failed to parse debug information for out/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.elf
llvm-objdump: warning: 'out/core/tee.elf': failed to parse debug information for out/core/tee.elf
```

Locally:
```
> clang --version #built using instructions from https://optee.readthedocs.io/en/latest/building/toolchains.html#llvm

clang version 9.0.1 (https://github.com/llvm/llvm-project.git c1a0a213378a458fbea1a5c77b315c7dce08fd05)
Target: x86_64-unknown-linux-gnu
Thread model: posix
```

```
> cd optee_os
> make PLATFORM=hikey-hikey960 CFG_ARM64_core=y COMPILER=clang

  CPP     out/arm-plat-hikey/ldelf/ldelf.lds
/tmp/ldelf-3acc92.s:1:1: error: unrecognized instruction mnemonic
OUTPUT_FORMAT("elf64-littleaarch64")
^
/tmp/ldelf-3acc92.s:2:1: error: unrecognized instruction mnemonic
OUTPUT_ARCH(aarch64)
^
/tmp/ldelf-3acc92.s:5:1: error: unrecognized instruction mnemonic
ENTRY(_ldelf_start)
^
/tmp/ldelf-3acc92.s:6:11: error: vector register expected
SECTIONS {
          ^
/tmp/ldelf-3acc92.s:7:2: error: invalid symbol redefinition
 .text : {
 ^
/tmp/ldelf-3acc92.s:7:10: error: unrecognized instruction mnemonic, did you mean: b?
 .text : {
         ^
/tmp/ldelf-3acc92.s:9:3: error: unexpected token at start of statement
  *(.text._ldelf_start)
  ^
/tmp/ldelf-3acc92.s:10:3: error: unexpected token at start of statement
  *(.text .text.*)
  ^
/tmp/ldelf-3acc92.s:11:3: error: unexpected token at start of statement
  *(.stub)
  ^
/tmp/ldelf-3acc92.s:12:3: error: unexpected token at start of statement
  *(.glue_7)
  ^
/tmp/ldelf-3acc92.s:13:3: error: unexpected token at start of statement
  *(.glue_7t)
  ^
/tmp/ldelf-3acc92.s:14:3: error: unexpected token at start of statement
  *(.gnu.linkonce.t.*)
  ^
/tmp/ldelf-3acc92.s:16:3: error: unexpected token at start of statement
  *(.vfp11_veneer)
  ^
/tmp/ldelf-3acc92.s:18:2: error: unrecognized instruction mnemonic, did you mean: b?
 }
 ^
/tmp/ldelf-3acc92.s:19:18: error: unknown token in expression
        .plt : { *(.plt) }
                 ^
/tmp/ldelf-3acc92.s:19:18: error: invalid operand
        .plt : { *(.plt) }
                 ^
/tmp/ldelf-3acc92.s:21:2: error: invalid symbol redefinition
 .eh_frame : { *(.eh_frame) }
 ^
/tmp/ldelf-3acc92.s:21:16: error: unknown token in expression
 .eh_frame : { *(.eh_frame) }
               ^
/tmp/ldelf-3acc92.s:21:16: error: invalid operand
 .eh_frame : { *(.eh_frame) }
               ^
/tmp/ldelf-3acc92.s:22:2: error: invalid symbol redefinition
 .rodata : {
 ^
/tmp/ldelf-3acc92.s:22:12: error: unrecognized instruction mnemonic, did you mean: b?
 .rodata : {
           ^
/tmp/ldelf-3acc92.s:23:3: error: unexpected token at start of statement
  *(.gnu.linkonce.r.*)
  ^
/tmp/ldelf-3acc92.s:24:3: error: unexpected token at start of statement
  *(.rodata .rodata.*)
  ^
/tmp/ldelf-3acc92.s:25:2: error: unrecognized instruction mnemonic, did you mean: b?
 }
 ^
/tmp/ldelf-3acc92.s:27:17: error: unknown token in expression
 .ARM.exidx : { *(.ARM.exidx* .gnu.linkonce.armexidx.*) }
                ^
/tmp/ldelf-3acc92.s:27:17: error: invalid operand
 .ARM.exidx : { *(.ARM.exidx* .gnu.linkonce.armexidx.*) }
                ^
/tmp/ldelf-3acc92.s:28:17: error: unknown token in expression
 .ARM.extab : { *(.ARM.extab*) }
                ^
/tmp/ldelf-3acc92.s:28:17: error: invalid operand
 .ARM.extab : { *(.ARM.extab*) }
                ^
/tmp/ldelf-3acc92.s:29:20: error: unknown token in expression
        .ctors : { *(.ctors) }
                   ^
/tmp/ldelf-3acc92.s:29:20: error: invalid operand
        .ctors : { *(.ctors) }
                   ^
/tmp/ldelf-3acc92.s:30:20: error: unknown token in expression
        .dtors : { *(.dtors) }
                   ^
/tmp/ldelf-3acc92.s:30:20: error: invalid operand
        .dtors : { *(.dtors) }
                   ^
/tmp/ldelf-3acc92.s:31:14: error: unknown token in expression
 .dynsym : { *(.dynsym) }
             ^
/tmp/ldelf-3acc92.s:31:14: error: invalid operand
 .dynsym : { *(.dynsym) }
             ^
/tmp/ldelf-3acc92.s:32:14: error: unknown token in expression
 .dynstr : { *(.dynstr) }
             ^
/tmp/ldelf-3acc92.s:32:14: error: invalid operand
 .dynstr : { *(.dynstr) }
             ^
/tmp/ldelf-3acc92.s:33:12: error: unknown token in expression
 .hash : { *(.hash) }
           ^
/tmp/ldelf-3acc92.s:33:12: error: invalid operand
 .hash : { *(.hash) }
           ^
/tmp/ldelf-3acc92.s:39:11: error: unexpected token
 . = ALIGN(8);
          ^
/tmp/ldelf-3acc92.s:41:15: error: unknown token in expression
 .rel.got : { *(.rel.got) }
              ^
/tmp/ldelf-3acc92.s:41:15: error: invalid operand
 .rel.got : { *(.rel.got) }
              ^
/tmp/ldelf-3acc92.s:42:16: error: unknown token in expression
 .rela.got : { *(.rela.got) }
               ^
/tmp/ldelf-3acc92.s:42:16: error: invalid operand
 .rela.got : { *(.rela.got) }
               ^
/tmp/ldelf-3acc92.s:43:16: error: unknown token in expression
 .rel.text : { *(.rel.text) *(.rel.gnu.linkonce.t*) }
               ^
/tmp/ldelf-3acc92.s:43:16: error: invalid operand
 .rel.text : { *(.rel.text) *(.rel.gnu.linkonce.t*) }
               ^
/tmp/ldelf-3acc92.s:44:17: error: unknown token in expression
 .rela.text : { *(.rela.text) *(.rela.gnu.linkonce.t*) }
                ^
/tmp/ldelf-3acc92.s:44:17: error: invalid operand
 .rela.text : { *(.rela.text) *(.rela.gnu.linkonce.t*) }
                ^
/tmp/ldelf-3acc92.s:45:16: error: unknown token in expression
 .rel.data : { *(.rel.data) *(.rel.gnu.linkonce.d*) }
               ^
/tmp/ldelf-3acc92.s:45:16: error: invalid operand
 .rel.data : { *(.rel.data) *(.rel.gnu.linkonce.d*) }
               ^
/tmp/ldelf-3acc92.s:46:17: error: unknown token in expression
 .rela.data : { *(.rela.data) *(.rela.gnu.linkonce.d*) }
                ^
/tmp/ldelf-3acc92.s:46:17: error: invalid operand
 .rela.data : { *(.rela.data) *(.rela.gnu.linkonce.d*) }
                ^
/tmp/ldelf-3acc92.s:47:18: error: unknown token in expression
 .rel.rodata : { *(.rel.rodata) *(.rel.gnu.linkonce.r*) }
                 ^
/tmp/ldelf-3acc92.s:47:18: error: invalid operand
 .rel.rodata : { *(.rel.rodata) *(.rel.gnu.linkonce.r*) }
                 ^
/tmp/ldelf-3acc92.s:48:19: error: unknown token in expression
 .rela.rodata : { *(.rela.rodata) *(.rela.gnu.linkonce.r*) }
                  ^
/tmp/ldelf-3acc92.s:48:19: error: invalid operand
 .rela.rodata : { *(.rela.rodata) *(.rela.gnu.linkonce.r*) }
                  ^
/tmp/ldelf-3acc92.s:49:15: error: unknown token in expression
 .rel.dyn : { *(.rel.dyn) }
              ^
/tmp/ldelf-3acc92.s:49:15: error: invalid operand
 .rel.dyn : { *(.rel.dyn) }
              ^
/tmp/ldelf-3acc92.s:50:16: error: unknown token in expression
 .rela.dyn : { *(.rela.dyn) }
               ^
/tmp/ldelf-3acc92.s:50:16: error: invalid operand
 .rela.dyn : { *(.rela.dyn) }
               ^
/tmp/ldelf-3acc92.s:51:17: error: unknown token in expression
 .rel.ctors : { *(.rel.ctors) }
                ^
/tmp/ldelf-3acc92.s:51:17: error: invalid operand
 .rel.ctors : { *(.rel.ctors) }
                ^
/tmp/ldelf-3acc92.s:52:18: error: unknown token in expression
 .rela.ctors : { *(.rela.ctors) }
                 ^
/tmp/ldelf-3acc92.s:52:18: error: invalid operand
 .rela.ctors : { *(.rela.ctors) }
                 ^
/tmp/ldelf-3acc92.s:53:17: error: unknown token in expression
 .rel.dtors : { *(.rel.dtors) }
                ^
/tmp/ldelf-3acc92.s:53:17: error: invalid operand
 .rel.dtors : { *(.rel.dtors) }
                ^
/tmp/ldelf-3acc92.s:54:18: error: unknown token in expression
 .rela.dtors : { *(.rela.dtors) }
                 ^
/tmp/ldelf-3acc92.s:54:18: error: invalid operand
 .rela.dtors : { *(.rela.dtors) }
                 ^
/tmp/ldelf-3acc92.s:55:16: error: unknown token in expression
 .rel.init : { *(.rel.init) }
               ^
/tmp/ldelf-3acc92.s:55:16: error: invalid operand
 .rel.init : { *(.rel.init) }
               ^
/tmp/ldelf-3acc92.s:56:17: error: unknown token in expression
 .rela.init : { *(.rela.init) }
                ^
/tmp/ldelf-3acc92.s:56:17: error: invalid operand
 .rela.init : { *(.rela.init) }
                ^
/tmp/ldelf-3acc92.s:57:16: error: unknown token in expression
 .rel.fini : { *(.rel.fini) }
               ^
/tmp/ldelf-3acc92.s:57:16: error: invalid operand
 .rel.fini : { *(.rel.fini) }
               ^
/tmp/ldelf-3acc92.s:58:17: error: unknown token in expression
 .rela.fini : { *(.rela.fini) }
                ^
/tmp/ldelf-3acc92.s:58:17: error: invalid operand
 .rela.fini : { *(.rela.fini) }
                ^
/tmp/ldelf-3acc92.s:59:15: error: unknown token in expression
 .rel.bss : { *(.rel.bss) }
              ^
/tmp/ldelf-3acc92.s:59:15: error: invalid operand
 .rel.bss : { *(.rel.bss) }
              ^
/tmp/ldelf-3acc92.s:60:16: error: unknown token in expression
 .rela.bss : { *(.rela.bss) }
               ^
/tmp/ldelf-3acc92.s:60:16: error: invalid operand
 .rela.bss : { *(.rela.bss) }
               ^
/tmp/ldelf-3acc92.s:61:15: error: unknown token in expression
 .rel.plt : { *(.rel.plt) }
              ^
/tmp/ldelf-3acc92.s:61:15: error: invalid operand
 .rel.plt : { *(.rel.plt) }
              ^
/tmp/ldelf-3acc92.s:62:16: error: unknown token in expression
 .rela.plt : { *(.rela.plt) }
               ^
/tmp/ldelf-3acc92.s:62:16: error: invalid operand
 .rela.plt : { *(.rela.plt) }
               ^
/tmp/ldelf-3acc92.s:66:11: error: unexpected token
 . = ALIGN(4096);
          ^
/tmp/ldelf-3acc92.s:68:15: error: unknown token in expression
 .dynamic : { *(.dynamic) }
              ^
/tmp/ldelf-3acc92.s:68:15: error: invalid operand
 .dynamic : { *(.dynamic) }
              ^
/tmp/ldelf-3acc92.s:69:11: error: unknown token in expression
 .got : { *(.got.plt) *(.got) }
          ^
/tmp/ldelf-3acc92.s:69:11: error: invalid operand
 .got : { *(.got.plt) *(.got) }
          ^
/tmp/ldelf-3acc92.s:71:2: error: invalid symbol redefinition
 .data : { *(.data .data.* .gnu.linkonce.d.*) }
 ^
/tmp/ldelf-3acc92.s:71:12: error: unknown token in expression
 .data : { *(.data .data.* .gnu.linkonce.d.*) }
           ^
/tmp/ldelf-3acc92.s:71:12: error: invalid operand
 .data : { *(.data .data.* .gnu.linkonce.d.*) }
           ^
/tmp/ldelf-3acc92.s:72:2: error: invalid symbol redefinition
 .bss : { *(.bss .bss.* .gnu.linkonce.b.* COMMON) }
 ^
/tmp/ldelf-3acc92.s:72:11: error: unknown token in expression
 .bss : { *(.bss .bss.* .gnu.linkonce.b.* COMMON) }
          ^
/tmp/ldelf-3acc92.s:72:11: error: invalid operand
 .bss : { *(.bss .bss.* .gnu.linkonce.b.* COMMON) }
          ^
/tmp/ldelf-3acc92.s:74:2: error: unexpected token at start of statement
 /DISCARD/ : { *(.interp) }
 ^
/tmp/ldelf-3acc92.s:75:1: error: unrecognized instruction mnemonic, did you mean: b?
}
^
ldelf/link.mk:53: recipe for target 'out/arm-plat-hikey/ldelf/ldelf.lds' failed
make: *** [out/arm-plat-hikey/ldelf/ldelf.lds] Error 1
```

**Comment by jforissier** [comment]
Hi @vchong ,

The two issues seem unrelated I would say.
- The first one (Shippable) seems pretty harmless, it happens when generating  `.dmp` files, debug stuff only, some time ago `llvm-objdump` wouldn't work at all (could not stand mixed Thumb and Arm code IIRC). I'll take a closer look though.
- As for your local build error, it's totally unexpected indeed. Can you try with `V=1` to see which command is executed? This step should simply run the C preprocessor on the linker script.

**Comment by vchong** [comment]
Hi @jforissier,

Thanks for the reply! So it was an user error after all. Because there's no `clang-cpp` in AOSP's version of `clang` [1], and in the manually built `clang-v9.0.1` tree, `clang-cpp` is just a symlink to `clang`, I changed `clang-cpp` to `clang` in `optee_os/mk/clang.mk` just to see if it'll build. It doesn't obviously per the errors reported above. As soon as I `clang` back to `clang-cpp`, the build is ok again, so apparently the symlink isn't just a symlink?

I've also tried to symlink `clang` to `clang-cpp` and use the prebuilt `clang` in [1] to build, but got similar errors as above. Note that there are various releases/versions of prebuilts in [1]. The ones I'd tried are `r353983c` (v9.0.3) and `r370808` (v10.0.1). Any ideas how to go about getting this to work?

Please note that `clang` from the current tip of llvm-project (sha 8c94d616, v11.0.0) doesn't work either, although there's only 1 error generated.
```
core/arch/arm/kernel/generic_entry_a64.S:426:2: error: changed section flags for .identity_map, expected: 0x6
 .section .identity_map
 ^
```

[1] https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/master

**Comment by jforissier** [comment]
> Hi @jforissier,
> 
> Thanks for the reply! So it was an user error after all. Because there's no `clang-cpp` in AOSP's version of `clang` [1], and in the manually built `clang-v9.0.1` tree, `clang-cpp` is just a symlink to `clang`, I changed `clang-cpp` to `clang` in `optee_os/mk/clang.mk` just to see if it'll build. It doesn't obviously per the errors reported above. As soon as I `clang` back to `clang-cpp`, the build is ok again, so apparently the symlink isn't just a symlink?

Not surprising at all. The symlink has nothing special, it is the compiler binary that uses `argv[0]` to detect what it is supposed to do. For an extreme illustration of this, try a QEMU build of OP-TEE and run `ls -l /bin` :wink: 

> I've also tried to symlink `clang` to `clang-cpp` and use the prebuilt `clang` in [1] to build, but got similar errors as above. Note that there are various releases/versions of prebuilts in [1]. The ones I'd tried are `r353983c` (v9.0.3) and `r370808` (v10.0.1). Any ideas how to go about getting this to work?

Well, `clang-cpp -> clang` should work IMO, unless the Andriod build is configured in a very special way, I have no idea.
 
> Please note that `clang` from the current tip of llvm-project (sha 8c94d616, v11.0.0) doesn't work either, although there's only 1 error generated.
> 
> ```
> core/arch/arm/kernel/generic_entry_a64.S:426:2: error: changed section flags for .identity_map, expected: 0x6
>  .section .identity_map
>  ^
> ```

Interesting, I will add this to my TODO list.

> [1] https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/master

**Comment by jforissier** [comment]
> Any ideas how to go about getting this to work?

I think `clang -E` should work:
```
CPP$(sm)        := $(ccache-cmd)clang --target=$(clang-target) -E
```

**Comment by vchong** [comment]
> I think clang -E should work:

It does! Thanks! Submitted https://github.com/OP-TEE/optee_os/pull/3769 for this.

**Comment by jforissier** [comment]
@vchong
> Please note that `clang` from the current tip of llvm-project (sha 8c94d616, v11.0.0) doesn't work either, although there's only 1 error generated.
> 
> ```
> core/arch/arm/kernel/generic_entry_a64.S:426:2: error: changed section flags for .identity_map, expected: 0x6
>  .section .identity_map
>  ^
> ```

Fixed in https://github.com/OP-TEE/optee_os/pull/3792.

**Comment by vchong** [comment]
> Fixed in #3792.

Confirmed. Thanks!
